pkgs/applications/virtualization/xen/4.18/4.18.nix

```diff
@@ -0,0 +1,192 @@
+{
+  lib,
+  callPackage,
+  fetchpatch,
+  fetchgit,
+  fetchFromGitHub,
+  ocaml-ng,
+
+  withInternalQEMU ? true,
+
+  withInternalTraditionalQEMU ? true,
+
+  withInternalSeaBIOS ? true,
+  withSeaBIOS ? !withInternalSeaBIOS,
+  seabios,
+
+  withInternalOVMF ? true,
+  withOVMF ? !withInternalOVMF,
+  OVMF,
+
+  withInternalIPXE ? true,
+  withIPXE ? !withInternalIPXE,
+  ipxe,
+
+  # QEMU
+  udev,
+  pciutils,
+  xorg,
+  SDL,
+  pixman,
+  acl,
+  glusterfs,
+  spice-protocol,
+  usbredir,
+  alsa-lib,
+  glib,
+  python311,
+
+  # OVMF
+  nasm,
+  bash,
+  ...
+}@args:
+
+assert withInternalSeaBIOS -> !withSeaBIOS;
+assert withInternalOVMF -> !withOVMF;
+assert withInternalIPXE -> !withIPXE;
+
+let
+  upstreamPatches = import ../patches.nix { inherit fetchpatch; };
+
+  qemuDeps = [
+    udev
+    pciutils
+    xorg.libX11
+    SDL
+    pixman
+    acl
+    glusterfs
+    spice-protocol
+    usbredir
+    alsa-lib
+    glib
+    python311
+  ];
+in
+
+callPackage (import ../generic.nix rec {
+  version = "4.18.2";
+  qemuVersion = "4.18.1";
+  seaBIOSVersion = "1.16.3";
+  ovmfVersion = "202305";
+
+  src = fetchgit {
+    url = "git://xenbits.xenproject.org/xen.git";
+    rev = "RELEASE-${version}";
+    hash = "sha256-pHCjj+Bcy4xQfB9xHU9fccFwVdP2DXrUhdszwGvrdmY=";
+  };
+
+  # Sources needed to build tools and firmwares.
+  prefetchedSources =
+    lib.attrsets.optionalAttrs withInternalQEMU {
+      qemu-xen = {
+        src = fetchgit {
+          url = "git://xenbits.xenproject.org/qemu-xen.git";
+          rev = "qemu-xen-${qemuVersion}";
+          hash = "sha256-BX+LXfNzwdUMALwwI1ZDW12dJ357oynjnrboLHREDGQ=";
+        };
+        buildInputs = qemuDeps;
+        postPatch = ''
+          substituteInPlace scripts/tracetool.py \
+            --replace-fail "/usr/bin/env python" "${python311}/bin/python"
+        '';
+        meta.description = "Xen's fork of upstream QEMU";
+      };
+    }
+    // lib.attrsets.optionalAttrs withInternalTraditionalQEMU {
+      qemu-xen-traditional = {
+        src = fetchgit {
+          url = "git://xenbits.xenproject.org/qemu-xen-traditional.git";
+          rev = "xen-${qemuVersion}";
+          hash = "sha256-R+I8QUFTeWzc/EIHibxfCoZa/wzcR59izVF4ciVshrU=";
+        };
+        buildInputs = qemuDeps;
+        postPatch = ''
+          substituteInPlace xen-hooks.mak \
+            --replace-fail /usr/include/pci ${pciutils}/include/pci
+        '';
+        meta.description = "Xen's fork of upstream QEMU that uses the traditional device model";
+      };
```

Comment:
Would it be feasable and good idea to implement a nixpkgs option that would add these patches to the current QEMU if Xen option is used? i feel that we should try reducing the amount of package forks that implement special features.

Reply:
It would be neither feasible, nor a good idea, nor in the scope of this PR.

1. `qemu-xen` isn't a collection of patches that can be used to patch vanilla QEMU. The trees have simply deviated too much, and Xen also pins their QEMU to specific versions in order to ensure compatibility with the hypervisor. `qemu-xen-traditional`, due to its age, is even more esoteric, as it's technically a really, really old version of QEMU that Xen still maintains to this day. (Though it is slowly being deprecated) Besides all that, `qemu-xen` isn't a package at all; it's just a build dependency. You can't find qemu-xen-traditional in the package search, as it isn't used for anything except for a full Xen build.

2. `qemu_xen`, with an underscore, already exists, and is what people who do not want a full Xen build should be using. It is comprised of vanilla QEMU, using QEMU's own Xen support. You may have noticed the with<Component> flags on the top of the `4.18.nix` file. Disabling the `withQEMU` and `withTraditionalQEMU` flags builds the hypervisor without the built-in QEMU, allowing it to reuse nixpkgs packages, such as `qemu_xen`.

Reply:
I'd recommend removing qemu-trad entirely.  This particular use of it is explicitly unsupported now.

Reply:
> I'd recommend removing qemu-trad entirely. This particular use of it is explicitly unsupported now.

So, always build with `--disable-traditional-qemu`?